helix-term/src/ui/editor.rs

```diff
@@ -695,6 +762,340 @@ impl EditorView {
+            );
+
+            // next node
+            context_area.y += 1;
+        }
+    }
+
+    fn get_context_paired_range(
+        query_match: &QueryMatch,
+        start_index: u32,
+        end_index: u32,
+        top_first_byte: usize,
+        last_scan_byte: usize,
+    ) -> Option<std::ops::Range<usize>> {
+        let end_nodes: Vec<_> = query_match.nodes_for_capture_index(end_index).collect();
+        query_match
+            .nodes_for_capture_index(start_index)
+            .flat_map(move |context| {
+                end_nodes.clone().into_iter().filter_map(move |it| {
+                    let start_range = context.byte_range();
+                    let end = it.start_byte();
+                    if start_range.contains(&end)
+                            && start_range.contains(&top_first_byte)
+                            && start_range.contains(&last_scan_byte)
+                            // only match @context.end nodes that aren't at the end of the line
+                            && context.start_position().row != it.start_position().row
+                    {
+                        Some(context.start_byte()..it.start_byte().saturating_sub(1))
+                    } else {
+                        None
+                    }
+                })
+                // in some cases, the start byte of a block is on the next line
+                // which causes to show the actual first line of content instead of
+                // the actual wanted "end of signature" line
+            })
+            .next()
+    }
+
+    /// Calculates the sticky nodes
+    fn calculate_sticky_nodes(
+        nodes: &Option<Vec<StickyNode>>,
+        doc: &Document,
+        view: &View,
+        config: &helix_view::editor::Config,
+    ) -> Option<Vec<StickyNode>> {
+        let syntax = doc.syntax()?;
+        let tree = syntax.tree();
+        let text = doc.text().slice(..);
+        let viewport = view.inner_area(doc);
+        let cursor_byte = text.char_to_byte(doc.selection(view.id).primary().cursor(text));
+
+        let visual_cursor_pos = view
+            .screen_coords_at_pos(doc, text, cursor_byte)
+            .unwrap_or_default()
+            .row as u16;
+
+        if visual_cursor_pos == 0 {
+            return None;
+        }
+
+        let anchor_line = text.char_to_line(view.offset.anchor);
+
+        let top_first_byte =
+            text.line_to_byte(anchor_line + nodes.as_ref().map_or(0, |nodes| nodes.len()));
+
+        let last_scan_byte = if config.sticky_context.follow_cursor {
+            cursor_byte
+        } else {
+            top_first_byte
+        };
+
+        // nothing has changed, so the cached result can be returned
+        if let Some(nodes) = nodes {
+            if nodes.iter().any(|node| view.offset.anchor == node.anchor) {
+                return Some(
+                    nodes
+                        .iter()
+                        .take(visual_cursor_pos as usize)
```

Comment:
shouldn't this be `visual_cursor_pos/3`?

Reply:
Ah yeah I got confused but the reason is that this doesn't take changihg view height into account. Basically the view could be very small and then you increase the size. Afterwards the cache would be incorrect. Is is the cache invalidated on resize?

I think you could solve this by setting some fixed maximum (like 100) and using that as a maximum for caching and then here you only take `min(view height/3, cursor)`